homeassistant/components/micropel/__init__.py

```diff
@@ -0,0 +1,151 @@
+"""Support for Micropel."""
+from abc import ABC
+import logging
+import typing
+from typing import ClassVar
+
+import voluptuous as vol
+
+from homeassistant.components.cover import DEVICE_CLASSES_SCHEMA
+from homeassistant.const import (
+    CONF_ADDRESS,
+    CONF_DEVICE_CLASS,
+    CONF_HOST,
+    CONF_NAME,
+    CONF_OFFSET,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_UNIQUE_ID,
+    CONF_UNIT_OF_MEASUREMENT,
+)
+from homeassistant.helpers import discovery
+import homeassistant.helpers.config_validation as cv
+
+from .const import (
+    CONF_COMMUNICATOR_TYPE,
+    CONF_CONNECTION_TCP,
+    CONF_DATA_TYPE,
+    CONF_PLC,
+    CONF_PRECISION,
+    CONF_REGISTER_TYPE,
+    CONF_SCALE,
+    DATA_TYPE_FLOAT,
+    DATA_TYPE_INT,
+    DOMAIN,
+    REGISTER_TYPE_LONG_WORD,
+    REGISTER_TYPE_WORD,
+    SupportedPlatforms,
+)
+from .micropel import MicropelModule
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class MicropelPlatformSchema(ABC):
+    """Voluptuous schema for Micropel platform entity configuration."""
+
+    PLATFORM_NAME: ClassVar[str]
+    ENTITY_SCHEMA: ClassVar[vol.Schema]
+
+    @classmethod
+    def platform_node(cls) -> typing.Dict[vol.Optional, vol.All]:
+        """Return a schema node for the platform."""
+        return {
+            vol.Optional(cls.PLATFORM_NAME): vol.All(
+                cv.ensure_list, [cls.ENTITY_SCHEMA]
+            )
+        }
+
+
+class SensorSchema(MicropelPlatformSchema):
+    """Voluptuous schema for Micropel sensors."""
+
+    PLATFORM_NAME = SupportedPlatforms.SENSOR.value
+
+    ENTITY_SCHEMA = vol.All(
+        vol.Schema(
```

Comment:
Is there a reason why you didn't implement the Config Flow? I am not familiar with Micropel PLC's, however it seems to mimick a device? Before I start a review, it would be good to understand if this is allowed due to ADR-0010.

https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md#decision

Reply:
Micropel integration is similar to integration like Modbus or KNX. There aren't specific devices, just memory addresses in PLC memory. The developer of the PLC program creates variables in PLC memory and the program works with them. This is the reason why I can't implement "Config Flow" and users have to configure HASS integration by YAML file. I expect that this kind of integration (include KNX and Modbus) will be able to do only experienced users with some technical background.

Reply:
@iMicknl Any news about this PR?

Reply:
Hi guys, @alexyao2015 @iMicknl. This PR has been open since 6 December. I would like to ask you if we can move forward or if there is anything missing from my side. I feel like we're stuck in a deadlock and we're not going anywhere.
Thank you in advance for your help.

Reply:
There are currently merge conflicts that need to be resolved.

Reply:
Bu there are CI failures 😜

Reply:
@spacegaier Fixed, I can see all green OKs.